In [1]:
import pandas as pd
import nltk
import vaderSentiment
import re
import string
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

from nltk.tokenize import sent_tokenize
from nltk.tokenize import word_tokenize
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\JZ\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

Read Data from CSV File:

In [2]:
path = r'C:\Users\JZ\Desktop\Data\JPMC_data_csv1.csv'

In [3]:
def load():
    pd.set_option("display.max_colwidth", None) #-1
    return pd.read_csv(path)

In [4]:
totalInfo = load().copy()

In [5]:
totalInfo

Date  \
0      10/24/2021   
1      10/24/2021   
2      10/24/2021   
3      10/24/2021   
4      10/24/2021   
...           ...   
26718   11/6/2021   
26719   11/6/2021   
26720   11/6/2021   
26721   11/6/2021   
26722   11/6/2021   

                                                                                                                                                        text  \
0                                                    $BBIG $AMC $GME $DWAC $BENE ✅📈📈📈 👉\n\n#DWAC #GME #BENE #BBIG #AMC https://t.co/PcMTemqch1   
1         RT @soFloTrading: SUNDAY GIVEAWAY TIME!\n\nOctober 18, 2021 - Weekly Trade Performance\n\nToday is the last day to lock in a FREE weekly trial!…   
2         We are going through the beginning stages of everything I’ve said in the last year about the economy. I’m no genius… https://t.co/DaBpCTtQh6   
3          $BBBY 🍋 candidate! the base is minimum. 🚀\n\n$GME $AMC $BB $NOK $NAKD $CCIV $BTC $ETH $DOGE $TEL $SHIB $ADA $XTZ… https://t.co/c1RCHDpDkF   
4      The largest mining company in the world\nGet Bonus 40% ⬇️\nhttps://t.co/kVp59h2td0\n\n$elon $doge $btc $tsla $amc $gme… https://t.co/I4ZfSukZaP   
...                                                                                                                                                      ...   
26718            @RobEducated $FB $MVRS Metaverse optionality isn’t really priced in at current ~22x P/E. I think I trust more Zuck… https://t.co/4w3U6cFthi   
26719           $FB $MRNA $NVAX $LAC $AAPL $NVDA\nMy calls , Weekly Options Gains :\n\n$FB 320p|Nov 05 | $1.55 to $3 &gt; 95% moved… https://t.co/c3k90lW7ye   
26720            @801010athlete I am viewing Thursday move was due to metaverse announcements from $FB and $MSFT. The N in CANSLIM.… https://t.co/oemrZW4Sp3   
26721            Good looking chart plays for the week. Enter long on PDH or a 5 minute squeeze. Nov. 12 Calls. \n$AAPL\n$AMZN\n$FB… https://t.co/bmNqvb2KsJ   
26722        @PatternProfits \n\n$FB\n\nHi Ben, can take a look at FB. It is good as long as it does not close below 326.04 for the… https://t.co/BPpkCMdeRp   

      favorite_count favorited retweeted lang quoted_status Stock_Name  \
0                  0     FALSE     FALSE   en           NaN        GME   
1                  0     FALSE     FALSE   en           NaN        GME   
2                  0     FALSE     FALSE   en           NaN        GME   
3                  0     FALSE     FALSE   en           NaN        GME   
4                  0     FALSE     FALSE   en           NaN        GME   
...              ...       ...       ...  ...           ...        ...   
26718              1     FALSE     FALSE   en           NaN         FB   
26719              1     FALSE     FALSE   en           NaN         FB   
26720              2     FALSE     FALSE   en           NaN         FB   
26721              1     FALSE     FALSE   en           NaN         FB   
26722              0     FALSE     FALSE   en           NaN         FB   

                                                          sentiment  \
0             {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}   
1      {'neg': 0.0, 'neu': 0.833, 'pos': 0.167, 'compound': 0.6825}   
2      {'neg': 0.095, 'neu': 0.905, 'pos': 0.0, 'compound': -0.296}   
3             {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}   
4      {'neg': 0.0, 'neu': 0.844, 'pos': 0.156, 'compound': 0.5423}   
...                                                             ...   
26718  {'neg': 0.0, 'neu': 0.852, 'pos': 0.148, 'compound': 0.5106}   
26719    {'neg': 0.0, 'neu': 0.906, 'pos': 0.094, 'compound': 0.34}   
26720         {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}   
26721  {'neg': 0.0, 'neu': 0.811, 'pos': 0.189, 'compound': 0.5994}   
26722  {'neg': 0.0, 'neu': 0.892, 'pos': 0.108, 'compound': 0.4404}   

      sentiment_class compound score  Stock Price  
0             neutral               0       1

In [6]:
#fill any NaN/invalid values
totalInfo = totalInfo.fillna(0)

In [7]:
totalInfo

Date  \
0      10/24/2021   
1      10/24/2021   
2      10/24/2021   
3      10/24/2021   
4      10/24/2021   
...           ...   
26718   11/6/2021   
26719   11/6/2021   
26720   11/6/2021   
26721   11/6/2021   
26722   11/6/2021   

                                                                                                                                                        text  \
0                                                    $BBIG $AMC $GME $DWAC $BENE ✅📈📈📈 👉\n\n#DWAC #GME #BENE #BBIG #AMC https://t.co/PcMTemqch1   
1         RT @soFloTrading: SUNDAY GIVEAWAY TIME!\n\nOctober 18, 2021 - Weekly Trade Performance\n\nToday is the last day to lock in a FREE weekly trial!…   
2         We are going through the beginning stages of everything I’ve said in the last year about the economy. I’m no genius… https://t.co/DaBpCTtQh6   
3          $BBBY 🍋 candidate! the base is minimum. 🚀\n\n$GME $AMC $BB $NOK $NAKD $CCIV $BTC $ETH $DOGE $TEL $SHIB $ADA $XTZ… https://t.co/c1RCHDpDkF   
4      The largest mining company in the world\nGet Bonus 40% ⬇️\nhttps://t.co/kVp59h2td0\n\n$elon $doge $btc $tsla $amc $gme… https://t.co/I4ZfSukZaP   
...                                                                                                                                                      ...   
26718            @RobEducated $FB $MVRS Metaverse optionality isn’t really priced in at current ~22x P/E. I think I trust more Zuck… https://t.co/4w3U6cFthi   
26719           $FB $MRNA $NVAX $LAC $AAPL $NVDA\nMy calls , Weekly Options Gains :\n\n$FB 320p|Nov 05 | $1.55 to $3 &gt; 95% moved… https://t.co/c3k90lW7ye   
26720            @801010athlete I am viewing Thursday move was due to metaverse announcements from $FB and $MSFT. The N in CANSLIM.… https://t.co/oemrZW4Sp3   
26721            Good looking chart plays for the week. Enter long on PDH or a 5 minute squeeze. Nov. 12 Calls. \n$AAPL\n$AMZN\n$FB… https://t.co/bmNqvb2KsJ   
26722        @PatternProfits \n\n$FB\n\nHi Ben, can take a look at FB. It is good as long as it does not close below 326.04 for the… https://t.co/BPpkCMdeRp   

      favorite_count favorited retweeted lang quoted_status Stock_Name  \
0                  0     FALSE     FALSE   en             0        GME   
1                  0     FALSE     FALSE   en             0        GME   
2                  0     FALSE     FALSE   en             0        GME   
3                  0     FALSE     FALSE   en             0        GME   
4                  0     FALSE     FALSE   en             0        GME   
...              ...       ...       ...  ...           ...        ...   
26718              1     FALSE     FALSE   en             0         FB   
26719              1     FALSE     FALSE   en             0         FB   
26720              2     FALSE     FALSE   en             0         FB   
26721              1     FALSE     FALSE   en             0         FB   
26722              0     FALSE     FALSE   en             0         FB   

                                                          sentiment  \
0             {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}   
1      {'neg': 0.0, 'neu': 0.833, 'pos': 0.167, 'compound': 0.6825}   
2      {'neg': 0.095, 'neu': 0.905, 'pos': 0.0, 'compound': -0.296}   
3             {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}   
4      {'neg': 0.0, 'neu': 0.844, 'pos': 0.156, 'compound': 0.5423}   
...                                                             ...   
26718  {'neg': 0.0, 'neu': 0.852, 'pos': 0.148, 'compound': 0.5106}   
26719    {'neg': 0.0, 'neu': 0.906, 'pos': 0.094, 'compound': 0.34}   
26720         {'neg': 0.0, 'neu': 1.0, 'pos': 0.0, 'compound': 0.0}   
26721  {'neg': 0.0, 'neu': 0.811, 'pos': 0.189, 'compound': 0.5994}   
26722  {'neg': 0.0, 'neu': 0.892, 'pos': 0.108, 'compound': 0.4404}   

      sentiment_class compound score  Stock Price  
0             neutral               0       1

In [8]:
#get rid of any noisy data (compound score is zero)
filter_noise = (totalInfo['compound score '] != '0')
totalInfo = totalInfo.loc[filter_noise]

In [9]:
totalInfo

Date  \
1      10/24/2021   
2      10/24/2021   
4      10/24/2021   
5      10/24/2021   
7      10/24/2021   
...           ...   
26717   11/6/2021   
26718   11/6/2021   
26719   11/6/2021   
26721   11/6/2021   
26722   11/6/2021   

                                                                                                                                                        text  \
1         RT @soFloTrading: SUNDAY GIVEAWAY TIME!\n\nOctober 18, 2021 - Weekly Trade Performance\n\nToday is the last day to lock in a FREE weekly trial!…   
2         We are going through the beginning stages of everything I’ve said in the last year about the economy. I’m no genius… https://t.co/DaBpCTtQh6   
4      The largest mining company in the world\nGet Bonus 40% ⬇️\nhttps://t.co/kVp59h2td0\n\n$elon $doge $btc $tsla $amc $gme… https://t.co/I4ZfSukZaP   
5                                                                        Crush them! Good luck to $AMC $GME we will soon unite $DWAC https://t.co/cA8Z1sYXtT   
7                                                    @WSBArmy I own $doge $amc $gme $shib $btc $eth and more. Love the power of momentum and community buzz.   
...                                                                                                                                                      ...   
26717         Making a fortune in $crypto\nBonus 40% \nhttps://t.co/KiE41CYre9\n\n$wish $amc $cei $gme $bbig $wdlf $kync\n$cei $gme… https://t.co/2Wws1VfWvp   
26718            @RobEducated $FB $MVRS Metaverse optionality isn’t really priced in at current ~22x P/E. I think I trust more Zuck… https://t.co/4w3U6cFthi   
26719           $FB $MRNA $NVAX $LAC $AAPL $NVDA\nMy calls , Weekly Options Gains :\n\n$FB 320p|Nov 05 | $1.55 to $3 &gt; 95% moved… https://t.co/c3k90lW7ye   
26721            Good looking chart plays for the week. Enter long on PDH or a 5 minute squeeze. Nov. 12 Calls. \n$AAPL\n$AMZN\n$FB… https://t.co/bmNqvb2KsJ   
26722        @PatternProfits \n\n$FB\n\nHi Ben, can take a look at FB. It is good as long as it does not close below 326.04 for the… https://t.co/BPpkCMdeRp   

      favorite_count favorited retweeted lang  \
1                  0     FALSE     FALSE   en   
2                  0     FALSE     FALSE   en   
4                  0     FALSE     FALSE   en   
5                  0     FALSE     FALSE   en   
7                  0     FALSE     FALSE   en   
...              ...       ...       ...  ...   
26717              0     FALSE     FALSE   en   
26718              1     FALSE     FALSE   en   
26719              1     FALSE     FALSE   en   
26721              1     FALSE     FALSE   en   
26722              0     FALSE     FALSE   en   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  

Find if sentence has ... in it

In [10]:
def remove_sentence(comment): #if text contains …, then it will be removed (… means truncated)
    print(comment.find('…'))
    if comment.find('…') > 0:
        #print("Remove")
        return 1
    else:
        #print("Don't remove")
        return -1
        

In [11]:
totalInfo['remove'] = totalInfo['text'].map(lambda x: '1' if remove_sentence(x) == 1 else '-1')

-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
30
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
38
-1
-1
-1
-1
-1
-1
-1
-1
-1
65
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
38
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-1
-

C:\Users\JZ\AppData\Local\Temp/ipykernel_18028/713790539.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  totalInfo['remove'] = totalInfo['text'].map(lambda x: '1' if remove_sentence(x) == 1 else '-1')


Filter out sentences that are truncated:

In [12]:
filter_truncated = totalInfo['remove'] == '-1'
full_comments_df = totalInfo.loc[filter_truncated]
full_comments_df

Date  \
1      10/24/2021   
2      10/24/2021   
4      10/24/2021   
5      10/24/2021   
7      10/24/2021   
...           ...   
26666   11/6/2021   
26667   11/6/2021   
26668   11/6/2021   
26677   11/6/2021   
26704   11/6/2021   

                                                                                                                                                        text  \
1         RT @soFloTrading: SUNDAY GIVEAWAY TIME!\n\nOctober 18, 2021 - Weekly Trade Performance\n\nToday is the last day to lock in a FREE weekly trial!…   
2         We are going through the beginning stages of everything I’ve said in the last year about the economy. I’m no genius… https://t.co/DaBpCTtQh6   
4      The largest mining company in the world\nGet Bonus 40% ⬇️\nhttps://t.co/kVp59h2td0\n\n$elon $doge $btc $tsla $amc $gme… https://t.co/I4ZfSukZaP   
5                                                                        Crush them! Good luck to $AMC $GME we will soon unite $DWAC https://t.co/cA8Z1sYXtT   
7                                                    @WSBArmy I own $doge $amc $gme $shib $btc $eth and more. Love the power of momentum and community buzz.   
...                                                                                                                                                      ...   
26666                                                                 RT @TrendSpider: $AMZN Rejected at the gap fill on the daily 😬 https://t.co/gfJ1pC0FdJ   
26667                                                                 RT @TrendSpider: $AMZN Rejected at the gap fill on the daily 😬 https://t.co/gfJ1pC0FdJ   
26668                                                                                  $AMZN Rejected at the gap fill on the daily 😬 https://t.co/gfJ1pC0FdJ   
26677                     Worried about inflation: \n\nThese 6 stocks might fight inflation better as per Morningstar.\n\n$BUD\n$KO\n$STZ\n$MSFT\n$FB\n$AMZN   
26704            Don't miss our next FREE OPTION TRADE.  Sign up for our Daily Free Trades at https://t.co/Y2nLzzBruA $NVDA $TSLA $GS $GILD $DKNG $GOOGL $FB   

      favorite_count favorited retweeted lang  \
1                  0     FALSE     FALSE   en   
2                  0     FALSE     FALSE   en   
4                  0     FALSE     FALSE   en   
5                  0     FALSE     FALSE   en   
7                  0     FALSE     FALSE   en   
...              ...       ...       ...  ...   
26666              0     FALSE     FALSE   en   
26667              0     FALSE     FALSE   en   
26668             38     FALSE     FALSE   en   
26677              2     FALSE     FALSE   en   
26704              0     FALSE     FALSE   en   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  

Before Filtering: 26722 rows

After Filtering: 15881 rows

60% of data remain

Modifed vaderSentiment Analyzer:

In [13]:
def one_comment(sentences):
    sentence_list = []
    stocks_mentioned = {} 
    seen = False
    within_comment = nltk.sent_tokenize(sentences) #makes a list of all sentences in comment
    paragraph_sentiment = 0.0

    for sentence in within_comment:
        seen = False
        analyzer = SentimentIntensityAnalyzer() #Create analyzer object
        analyzer.polarity_scores(within_comment)
        vs = analyzer.polarity_scores(sentence) # sentiment of sentence
        words = nltk.word_tokenize(sentence) # makes a list of all the words within a sentence
        for word in words: #iterates through each word in sentence
            if word in all_stocks: #checks if stock name is a word
                if word not in stocks_mentioned: #adds word to dictionary
                    stocks_mentioned[word] = vs['compound'] #sets value of that stock sentiment to vs['compound']
                    seen = True  # Ex: "$AMC best stock $AMC" <--- if repetition in one sentence (we don't want to add to value)
                elif word in stocks_mentioned: #increases compound score
                    if seen == False: # Ex: "$AMC best stock $AMC" <--- repetition in one sentence
                        stocks_mentioned[word] += vs['compound'] 
        paragraph_sentiment += vs['compound']

    #print(within_comment, stocks_mentioned, paragraph_sentiment)
    return stocks_mentioned

In [14]:
"""### Sentiment Analyzer"""

def company_scores(sentence): #returns a dictionary with company names in interest with compound score 
    sentiment_dict = one_comment(sentence)
    return sentiment_dict


def apply_vadersentiment(df):
    df['modified']=df['text'].apply(lambda x: company_scores(x))
    #df['sentiment_class']=df['sentiment'].apply(lambda x: 'positive' if x['compound']>0.5 else ('negative' if x['compound']<-0.5 else 'neutral'))
    #df['compound score']=df['sentiment'].apply(lambda x: x['compound'])
    return df

Obtain all stocks that we are interested in:

In [15]:
stocks = totalInfo['Stock_Name'].to_numpy()
stocks

all_stocks = set(stocks)
all_stocks.remove('Stock_Name')

all_stocks

{'AMC', 'AMZN', 'BAC', 'FB', 'GME', 'JPM', 'NOK', 'SBUX', 'TSLA'}

In [16]:
apply_vadersentiment(full_comments_df)

C:\Users\JZ\AppData\Local\Temp/ipykernel_18028/14248610.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['modified']=df['text'].apply(lambda x: company_scores(x))


Date  \
1      10/24/2021   
2      10/24/2021   
4      10/24/2021   
5      10/24/2021   
7      10/24/2021   
...           ...   
26666   11/6/2021   
26667   11/6/2021   
26668   11/6/2021   
26677   11/6/2021   
26704   11/6/2021   

                                                                                                                                                        text  \
1         RT @soFloTrading: SUNDAY GIVEAWAY TIME!\n\nOctober 18, 2021 - Weekly Trade Performance\n\nToday is the last day to lock in a FREE weekly trial!…   
2         We are going through the beginning stages of everything I’ve said in the last year about the economy. I’m no genius… https://t.co/DaBpCTtQh6   
4      The largest mining company in the world\nGet Bonus 40% ⬇️\nhttps://t.co/kVp59h2td0\n\n$elon $doge $btc $tsla $amc $gme… https://t.co/I4ZfSukZaP   
5                                                                        Crush them! Good luck to $AMC $GME we will soon unite $DWAC https://t.co/cA8Z1sYXtT   
7                                                    @WSBArmy I own $doge $amc $gme $shib $btc $eth and more. Love the power of momentum and community buzz.   
...                                                                                                                                                      ...   
26666                                                                 RT @TrendSpider: $AMZN Rejected at the gap fill on the daily 😬 https://t.co/gfJ1pC0FdJ   
26667                                                                 RT @TrendSpider: $AMZN Rejected at the gap fill on the daily 😬 https://t.co/gfJ1pC0FdJ   
26668                                                                                  $AMZN Rejected at the gap fill on the daily 😬 https://t.co/gfJ1pC0FdJ   
26677                     Worried about inflation: \n\nThese 6 stocks might fight inflation better as per Morningstar.\n\n$BUD\n$KO\n$STZ\n$MSFT\n$FB\n$AMZN   
26704            Don't miss our next FREE OPTION TRADE.  Sign up for our Daily Free Trades at https://t.co/Y2nLzzBruA $NVDA $TSLA $GS $GILD $DKNG $GOOGL $FB   

      favorite_count favorited retweeted lang  \
1                  0     FALSE     FALSE   en   
2                  0     FALSE     FALSE   en   
4                  0     FALSE     FALSE   en   
5                  0     FALSE     FALSE   en   
7                  0     FALSE     FALSE   en   
...              ...       ...       ...  ...   
26666              0     FALSE     FALSE   en   
26667              0     FALSE     FALSE   en   
26668             38     FALSE     FALSE   en   
26677              2     FALSE     FALSE   en   
26704              0     FALSE     FALSE   en   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  